### PR TITLE
Add a checker to confirm physical optimizer rules will keep the physical plan schema immutable

### LINKED ISSUE #4299: Add a checker to confirm physical optimizer rules will keep the physical plan schema immutable
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #4299.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@jackwener @andygrove @alamb

File: datafusion/core/src/physical_optimizer/optimizer.rs [hunk 1]
```diff
@@ -36,4 +36,10 @@ pub trait PhysicalOptimizerRule {
 
     /// A human readable name for this optimizer rule
     fn name(&self) -> &str;
+
+    /// A flag to indicate whether the physical planner should valid the rule will not
+    /// change the schema of the plan after the rewriting.
+    /// Some of the optimization rules might change the nullable properties of the schema
+    /// and should disable the schema check.
+    fn schema_check(&self) -> bool;
```
Comment (jackwener) [inline datafusion/core/src/physical_optimizer/optimizer.rs:44]: 👍

**Review by jackwener — APPROVED** [review]
Nice job.

**Review by alamb — APPROVED** [review]
Thanks @mingmwang

**Pull request merged**